[Swift bridging] Consistently use correctly-defined Swift(U)Int for bridging

DougGregor opened the pull request:
The `SwiftInt` and `SwiftUInt` types in the compiler provide `C` equivalents to Swift's `Int` and `UInt`. Consistently use these types in the C/Swift bridging layer, to avoid painful truncations and data-size mismatches.

This is harder than it should be. The canonical C types are `ptrdiff_t` and `size_t`, but we're in the depths of the compiler we can't include the C library headers that provide them because they introduce cyclic module dependencies. Sigh.

SwiftShims has some logic to compute these types. However, SwiftShims is part of the Swift runtime, not the compiler, so those headers cannot be included here. So, we clone the logic and simplify it somewhat for our use case.

This fixes truncation issues on Windows, where the uses of `unsigned long` and `long` for Swift(U)Int are incorrect.

DougGregor committed:
[Swift bridging] Use Swift(U)Int consistently for bridging

DougGregor committed:
Define the C++ Swift(U)Int types using the same logic as SwiftShims

The C type tha corresponds to Swift's pointer-sized `Int` and `UInt`
types varies from one platform to the next. The canonical C types are
`ptrdiff_t` and `size_t`, but we're in the depths of the compiler we
can't include the C library headers that provide them because they
introduce cyclic module dependencies. Sigh.

SwiftShims has some logic to compute these types. However, SwiftShims
is part of the Swift runtime, not the compiler, so those headers
cannot be included here. So, we clone the logic and simplify it
somewhat for our use case.

This fixes truncation issues on Windows, where the uses of
`unsigned long` and `long` for Swift(U)Int are incorrect.

DougGregor commented:
@swift-ci please smoke test

compnerd commented on include/swift/Basic/CBasicBridging.h:
```diff
@@ -21,6 +21,64 @@
+typedef __swiftc_uint32_t SwiftUInt;
+#elif defined(_WIN64)
+typedef __swiftc_int64_t SwiftInt;
+typedef __swiftc_uint64_t SwiftUInt;
```
This is inverted ... `_WIN64` will have `_WIN32` defined IIRC, so we will always create 32-bit typedefs.

DougGregor replied:
Oh no! This comes out of SwiftShims. I can fix in both places, but I don't have a good way to test the fix

DougGregor replied:
Fixed in https://github.com/apple/swift/pull/68278/commits/592752741917c0ff76e382646f6d3e8a0bdd30c1, thank you so much!

DougGregor committed:
Switch order of _WIN64/_WIN32 checks in computation of C types for Int/UInt

64-bit Windows defines both _WIN64 and _WIN32, so the logic here would
always end up defining 32-bit C types for Swift's `Int` and `UInt`.
Fix the ordering to check for 64-bit first, then 32-bit second.

Note that the SwiftShims version of this code has always been wrong,
but it's completely benign because SwiftShims is only used in the
Swift runtime itself, which is built with Clang (on all platforms),
and doesn't need to go through this code path. Still, we fix it in both
places, so we don't get a nasty surprise if the SwiftShims version of
the header later gets included in a non-Clang C++ compiler.

DougGregor commented:
@swift-ci please smoke test

compnerd reviewed (approved):
Awesome!  Thanks again @DougGregor!

DougGregor committed:
Fix handling of Swift(U)Int for Windows

DougGregor commented:
@swift-ci please smoke test

compnerd commented on include/swift/Basic/CBasicBridging.h:
```diff
@@ -21,6 +21,69 @@
+typedef __swiftc_intn_t(__INTPTR_WIDTH__) SwiftInt;
+typedef __swiftc_uintn_t(__INTPTR_WIDTH__) SwiftUInt;
+#endif
+#endif
```
```suggestion
```

compnerd commented on include/swift/Basic/CBasicBridging.h:
```diff
@@ -21,6 +21,69 @@
+typedef __UINT64_TYPE__ __swiftc_uint64_t;
+#else
+typedef unsigned __INT64_TYPE__ __swiftc_uint64_t;
+#endif
```
```suggestion
#endif
#endif
```

DougGregor committed:
Apply suggestions from code review

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

DougGregor commented:
@swift-ci please smoke test

compnerd commented on include/swift/Basic/CBasicBridging.h:
```diff
@@ -21,6 +21,69 @@
+#else
+typedef unsigned __INT64_TYPE__ __swiftc_uint64_t;
+#endif
+#endif
```
```suggestion
```

compnerd commented on include/swift/Basic/CBasicBridging.h:
```diff
@@ -21,6 +21,69 @@
+typedef __UINT32_TYPE__ __swiftc_uint32_t;
+#else
+typedef unsigned __INT32_TYPE__ __swiftc_uint32_t;
+#endif
```
```suggestion
#endif
#endif
```

DougGregor committed:
Apply suggestions from code review

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

DougGregor commented:
@swift-ci please smoke test

compnerd committed:
ASTGen: adjust for type changes

The type definitions do not uniformly import as `Swift.Int` and `Swift.UInt`.  Add some casts to accommodate the type conversions.

compnerd commented:
@swift-ci please smoke test
